generic/tests/pvpanic.py

```diff
@@ -0,0 +1,139 @@
+    :vm: target vm
+    :session: session created by loggin the vm
+    :enable: to setup or restore the test environment
+    """
+    timeout = params.get("timeout", 360)
+    if enable:
+        kdump_cmd = params.get("kdump_disable_cmd")
+        nmi_panic_cmd = params.get("nmi_enable_cmd")
+        registry_cmd = params.get("NMICrashDump_enable_cmd")
+    else:
+        kdump_cmd = params.get("kdump_enable_cmd")
+        nmi_panic_cmd = params.get("nmi_disable_cmd")
+        registry_cmd = params.get("NMICrashDump_disable_cmd")
+
+    switch_cmds = []
+    if params.get("os_type") == "linux":
+        # switch kdump service and unknown_nmi_panic
+        switch_cmds.append(kdump_cmd)
+        switch_cmds.append(nmi_panic_cmd)
+    else:
+        # modify the register for windows
+        switch_cmds.append(registry_cmd)
+
+    for cmd in switch_cmds:
+        status, output = session.cmd_status_output(cmd, timeout=60)
+        if status:
+            test.error("Command '%s' failed, status: %s, output: %s" %
+                       (cmd, status, output))
+    if params.get("os_type") == "windows":
+        session = vm.reboot(session, timeout=timeout)
+
+
+def check_qmp_events(vm, event_name, timeout=600):
+    """
+    Check whether certain qmp event appeared in vm.
+
+    :param vm: target virtual machine
+    :param event_name: target event name, such as 'GUEST_PANICKED'
+    :param timeout: check time
+    """
+
+    end_time = time.time() + timeout
+    qmp_monitors = vm.get_monitors_by_type("qmp")
+    qmp_num = len(qmp_monitors)
+    logging.info("Try to get qmp events %s in %s seconds!" %
+                 (event_name, timeout))
+    while time.time() < end_time:
+        for monitor in qmp_monitors:
+            event = monitor.get_event(event_name)
+            if event:
+                logging.info("Receive qmp %s event notification",
+                             event_name)
+                qmp_num -= 1
+                qmp_monitors.remove(monitor)
+        if qmp_num == 0:
+            return True
+        time.sleep(5)
+    return False
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    pvpanic test
+
+    1) Log into the guest
+    2) Check if the driver is installed and verified (only for win)
+    3) Stop kdump service and modify unknown_nmi_panic(for linux)
+       or modify register value(for win)
+    4) Trigger a crash by nmi
+    5) Check the event in qmp
+    6) Check guest status
+
+    :param test: kvm test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    timeout = float(params.get("timeout", 360))
+    vm = env.get_vm(params["main_vm"])
+    event_check = params.get("event_check", "GUEST_PANICKED")
+    guest_status = params.get("guest_status", "guest-panicked")
+    vm.verify_alive()
+    session = vm.wait_for_login(timeout=timeout)
+    error_context.context("Boot guest with pvpanic device", logging.info)
```

Comment:
It's better to move this info  before env.get_vm() line.

Reply:
@xiagao updated according to your advices
Thanks